### PR TITLE
[WIP] Fixes #4108 | Add logic to check if provided storage class exist

### PR DESCRIPTION
**Pull request by chiukapoor**
## Description

- [x] Add logic to check if provided storage class exists
- [ ] Update the RBAC to provide access to storageclass
- [x] Check if the operator has access to storageclass
- [ ] Update test case

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

Add logic to check if provided storage class exists along with the required permission and required fallback checks

```release-note

```

### CODE REVIEW

File: cmd/operator/main.go [hunk 1]
```diff
@@ -271,6 +272,25 @@ func run() int {
...
+		ctx,
+		logger,
+		cc,
+		allowedNamespaces,
+		map[string][]string{
+			storageclassv1.GroupName: {"get"},
+		},
+		storageclassv1.SchemeGroupVersion.String(),
+		storageclassv1.GroupName,
+	)
```
Comment (ApsTomar) [inline cmd/operator/main.go]: This looks like an overkill, the arguments are pretty small to occupy one line per arg. Is this the way, they're doing it everywhere else in the code? Better way would be to club smaller items in single line and break the line where it gets too long.

File: pkg/alertmanager/operator.go [hunk 2]
```diff
@@ -654,6 +658,21 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 
 	level.Info(logger).Log("msg", "sync alertmanager")
 
+	if c.canReadStorageClass {
+		if !reflect.DeepEqual(am.Spec.Storage.VolumeClaimTemplate, monitoringv1.EmbeddedPersistentVolumeClaim{}) {
+			storageClassName := am.Spec.Storage.VolumeClaimTemplate.Spec.StorageClassName
+			if storageClassName != nil && *storageClassName != "" {
+				_, err := c.kclient.StorageV1().StorageClasses().Get(ctx, *storageClassName, metav1.GetOptions{})
+				if err != nil {
```
Comment (ApsTomar) [inline pkg/alertmanager/operator.go]: That's a lot of nesting for an if-block. See if you can refactor it to make the code more readable. If not, then atleast put some comments explaining the 2nd and 3rd if conditions (line 662 and 664).

**Review by ApsTomar — CHANGES_REQUESTED** [review]
Requested some changes in the code.

File: cmd/operator/main.go (hunk 1 above)
Comment (chiukapoor) [reply cmd/operator/main.go]: Yes, that is how they are doing it when calling checkPrequisites everywhere in the file

File: pkg/alertmanager/operator.go (hunk 2 above)
Comment (chiukapoor) [reply pkg/alertmanager/operator.go]: Sure will refactor it

Comment (chiukapoor) [reply pkg/alertmanager/operator.go]: Updated

File: cmd/operator/main.go (hunk 1 above)
Comment (chiukapoor) [reply cmd/operator/main.go]: https://github.com/infracloudio/prometheus-operator/blob/b0aa730488f093be9fa39409d49d9cd8a67edc77/cmd/operator/main.go#L274-L289

**Comment by chiukapoor** [comment]
Removed the RBAC as the same seems to be updated using jsonnet file. They are not using the general RBAC policy create using kubebuilder.

**Pull request merged**